### PR TITLE
layoutCallback should properly cancel if unlayoutCallback was called

### PR DESCRIPTION
**Pull request by bradfrizzell**
Need to make layoutCallback handle the case where unlayoutCallback was called after layoutCallback had dropped the thread. 

Work in progress, didn't update jsdoc or add any tests yet. Just want to check if this is basically the right approach.

### CODE REVIEW

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 1]
```diff
@@ -830,6 +830,13 @@ export class AmpA4A extends AMP.BaseElement {
 
   /** @override */
   layoutCallback() {
+    // Shorthand for: reject promise if current promise chain is out of date.
+    const checkStillCurrent = promiseId => {
+      if (promiseId != this.promiseId_) {
+        throw cancellation();
+      }
+    };
+    const promiseId = this.promiseId_;
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: I would do:
    `const promiseId = this.promiseId_;`
     `const checkStillCurrent = () => {`
     `  if (promiseId != this.promiseId_) {`
     `    throw cancellation();`
     `  }`
   `};`
Then you shouldn't need to pass down promiseId

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 2]
```diff
@@ -859,17 +867,18 @@ export class AmpA4A extends AMP.BaseElement {
...
         // Non-AMP creative case, will verify ad url existence.
-        return this.renderNonAmpCreative_();
+        return this.renderNonAmpCreative_(checkStillCurrent, promiseId);
       }
       // Must be an AMP creative.
-      return this.renderAmpCreative_(creativeMetaData)
+      return this.renderAmpCreative_(creativeMetaData, checkStillCurrent,
+                                    promiseId)
         .catch(err => {
           // Failed to render via AMP creative path so fallback to non-AMP
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Add checkStillCurrent here

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 1 above)
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: d'oh, will fix

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -859,18 +867,19 @@ export class AmpA4A extends AMP.BaseElement {
       }
       if (!creativeMetaData) {
         // Non-AMP creative case, will verify ad url existence.
-        return this.renderNonAmpCreative_();
+        return this.renderNonAmpCreative_(checkStillCurrent);
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: I don't think you actually need to pass checkStillCurrent around if you believe you have checked the current promiseId against the global prior to any promise releases.  You should just be able to get the current promiseId prior to promise and compare.

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 4]
```diff
@@ -377,6 +377,60 @@ describe('amp-a4a', () => {
...
+      let promiseResolver;
+      a4a.adPromise_ = new Promise(resolver => {
+        promiseResolver = resolver;
+      });
+      const layoutCallbackPromise = a4a.layoutCallback();
+      a4a.unlayoutCallback();
+      const renderNonAmpCreativeSpy = sinon.spy(AmpA4A.prototype, 'renderNonAmpCreative_');
+      promiseResolver();
+      layoutCallbackPromise.catch(err => {
+        console.log(err);
```
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: done

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: done

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 5]
```diff
@@ -377,6 +377,60 @@ describe('amp-a4a', () => {
...
+      let promiseResolver;
+      a4a.renderAmpCreative_ = new Promise(resolver => {
+        promiseResolver = resolver;
+      });
+      const layoutCallbackPromise = a4a.layoutCallback();
+      a4a.unlayoutCallback();
+
+      promiseResolver();
+      layoutCallbackPromise.catch(err => {
+        expect(err).to.be.ok;
```
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: done

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js (hunk 4 above)
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: done

Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: done

**Comment by bradfrizzell** [comment]
ptal

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 6]
```diff
@@ -861,16 +865,20 @@ export class AmpA4A extends AMP.BaseElement {
         // Non-AMP creative case, will verify ad url existence.
         return this.renderNonAmpCreative_();
       }
+      promiseId = this.promiseId_;
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Why is this necessary?  By definition it should still be the same value otherwise we should have cancelled.  Make promiseId a const

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 7]
```diff
@@ -1190,7 +1198,7 @@ export class AmpA4A extends AMP.BaseElement {
         }
       });
     }
-
+    let promiseId = this.promiseId_;
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Same here, make promiseId a const

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 8]
```diff
@@ -1214,9 +1225,14 @@ export class AmpA4A extends AMP.BaseElement {
               this.getAmpDoc(), friendlyIframeEmbed.win);
           // Bubble phase click handlers on the ad.
           this.registerAlpHandler_(friendlyIframeEmbed.win);
+          promiseId = this.promiseId_;
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Remove

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 9]
```diff
@@ -1228,11 +1244,15 @@ export class AmpA4A extends AMP.BaseElement {
             dev().error(TAG, this.element.getAttribute('type'),
                 'Error executing onCreativeRender', err);
           })(true);
+          promiseId = this.promiseId_;
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: remove

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 10]
```diff
@@ -1276,28 +1338,29 @@ describe('amp-a4a', () => {
       });
     });
     it('should render correctly', () => {
-      return a4a.renderAmpCreative_(metaData).then(() => {
+      return a4a.renderAmpCreative_(metaData, checkStillCurrent)
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: These are outdated given renderAmpCreative_ no longer takes checkStillCurrent right?

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 8 above)
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: done

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 7 above)
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: done

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 6 above)
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: done

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 9 above)
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: done

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js (hunk 10 above)
Comment (bradfrizzell) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: whoops, yes. done

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 11]
```diff
@@ -1276,24 +1337,25 @@ describe('amp-a4a', () => {
       });
     });
     it('should render correctly', () => {
-      return a4a.renderAmpCreative_(metaData).then(() => {
+      return a4a.renderAmpCreative_(metaData)
+          .then(() => {
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: Revert these changes given they are just formatting?

**Comment by bradfrizzell** [comment]
# **_friendly ping_**
